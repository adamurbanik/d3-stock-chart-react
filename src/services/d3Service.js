import * as d3 from 'd3';

class D3Service {
  constructor() {
    this.parseDate = d3.timeParse('%Y-%m-%d');

    this.stockNumber = 1;
    this.chartNumber = 1;
  }

  removeAll() {
    this.removeStock();
    this.removeChart();
  }

  prepareStock() {
    this.svgStock = d3.select('.svg-stock');
    const marginStock = { top: 60, right: 20, bottom: 100, left: 40 };
    this.widthStock =
      +this.svgStock.attr('width') - marginStock.left - marginStock.right;
    this.heightStock =
      +this.svgStock.attr('height') - marginStock.top - marginStock.bottom;

    this.xStock = d3.scaleTime().range([0, this.widthStock]);
    this.yStock = d3.scaleLinear().range([this.heightStock, 0]);

    this.xAxisStock = d3.axisBottom(this.xStock);
    this.yAxisStock = d3.axisLeft(this.yStock);

    this.brush = d3
      .brushX()
      .extent([
        [0, 0],
        [this.widthStock, this.heightStock]
      ])
      .on('brush end', this.brushed.bind(this));

    d3.area()
      .curve(d3.curveMonotoneX)
      .x(d => this.xStock(d.date))
      .y0(this.heightStock)
      .y1(d => this.yStock(d.high));

    this.valueline = d3
      .line()
      .x(d => this.xStock(d.date))
      .y(d => this.yStock(d.high));

    this.context = this.svgStock
      .append('g')
      .attr('class', 'context')
      .attr(
        'transform',
        'translate(' + marginStock.left + ',' + marginStock.top + ')'
      );
  }

  removeStock() {
    if (this.svgStock) {
      d3.select('.svg-stock')
        .selectAll('*')
        .remove();
    }
  }

  applyStockData(data) {
    this.removeStock();
    this.prepareStock();

    data.forEach(d => {
      d.date = this.parseDate(d[0]);
      d.high = +d[2];
      d.low = +d[3];
    });

    this.xStock.domain(
      d3.extent(data, function(d) {
        return d.date;
      })
    );
    this.yStock.domain([
      d3.min(data, function(d) {
        return d.low;
      }),
      d3.max(data, function(d) {
        return d.high;
      })
    ]);

    let classNames = `line line${this.stockNumber}`;
    this.context
      .append('path')
      .datum(data)
      .attr('class', classNames)
      .attr('d', this.valueline);

    this.context
      .append('g')
      .attr('class', 'axis axis--x')
      .attr('transform', 'translate(0,' + this.heightStock + ')')
      .call(this.xAxisStock);

    this.context
      .append('g')
      .attr('class', 'axis axis--y')
      .call(this.yAxisStock);
  }

  prepareChart() {
    this.svgChart = d3.select('.svg-chart');

    this.margin = { top: 20, right: 20, bottom: 110, left: 40 };
    this.width =
      +this.svgChart.attr('width') - this.margin.left - this.margin.right;
    this.height =
      +this.svgChart.attr('height') - this.margin.top - this.margin.bottom;

    this.x = d3.scaleTime().range([0, this.width]);
    this.y = d3.scaleLinear().range([this.height, 0]);

    this.xAxis = d3.axisBottom(this.x);
    this.yAxis = d3.axisLeft(this.y);

    this.zoom = d3
      .zoom()
      .scaleExtent([1, Infinity])
      .translateExtent([
        [0, 0],
        [this.width, this.height]
      ])
      .extent([
        [0, 0],
        [this.width, this.height]
      ])
      .on('zoom', this.zoomed.bind(this));

    this.area = d3
      .area()
      .curve(d3.curveMonotoneX)
      .x(d => this.x(d.date))
      .y0(this.height)
      .y1(d => this.y(d.high));

    this.svgChart
      .append('defs')
      .append('clipPath')
      .attr('id', 'clip')
      .append('rect')
      .attr('width', this.width)
      .attr('height', this.height);

    this.focus = this.svgChart
      .append('g')
      .attr('class', 'focus')
      .attr(
        'transform',
        'translate(' + this.margin.left + ',' + this.margin.top + ')'
      );
  }

  removeChart() {
    if (this.svgChart) {
      d3.select('.svg-chart')
        .selectAll('*')
        .remove();
    }
  }

  applyChartData(data) {
    this.removeChart();
    this.prepareChart();

    this.x.domain(
      d3.extent(data, function(d) {
        return d.date;
      })
    );
    this.y.domain([
      d3.min(data, function(d) {
        return d.low;
      }),
      d3.max(data, function(d) {
        return d.high;
      })
    ]);

    this.focus
      .append('path')
      .datum(data)
      .attr('class', `area${this.chartNumber}`)
      .attr('d', this.area);

    this.focus
      .append('g')
      .attr('class', 'axis axis--x')
      .attr('transform', 'translate(0,' + this.height + ')')
      .call(this.xAxis);

    this.focus
      .append('g')
      .attr('class', 'axis axis--y')
      .call(this.yAxis);

    this.svgChart
      .append('rect')
      .attr('class', 'zoom')
      .attr('width', this.width)
      .attr('height', this.height)
      .attr(
        'transform',
        'translate(' + this.margin.left + ',' + this.margin.top + ')'
      )
      .call(this.zoom);

    this.context
      .append('g')
      .attr('class', 'brush')
      .call(this.brush)
      .call(this.brush.move, this.xStock.range());
  }

  brushed() {
    if (d3.event.sourceEvent && d3.event.sourceEvent.type === 'zoom') return;
    if (!this.focus) return;
    var s = d3.event.selection || this.xStock.range();

    this.x.domain(s.map(this.xStock.invert, this.xStock));
    this.focus.select('.area1').attr('d', this.area);
    this.focus.select('.axis--x').call(this.xAxis);
    this.svgChart
      .select('.zoom')
      .call(
        this.zoom.transform,
        d3.zoomIdentity.scale(this.width / (s[1] - s[0])).translate(-s[0], 0)
      );
  }

  zoomed() {
    if (d3.event.sourceEvent && d3.event.sourceEvent.type === 'brush') return;
    var t = d3.event.transform;

    this.x.domain(t.rescaleX(this.xStock).domain());

    this.focus.select('.area1').attr('d', this.area);
    this.focus.select('.axis--x').call(this.xAxis);
    this.context
      .select('.brush')
      .call(this.brush.move, this.x.range().map(t.invertX, t));
  }
}

export default D3Service;
