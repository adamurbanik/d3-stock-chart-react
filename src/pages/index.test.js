import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Stock from './index';
import d3Service from '../services/d3Service';

jest.mock('../services/d3Service');

describe('<Stock />', () => {
  let dataset;
  beforeAll(() => {
    d3Service.applyStockData = jest.fn();
    d3Service.applyChartData = jest.fn();
    dataset = 'AAPL';
  });

  test('renders correctly', () => {
    const wrapper = shallow(<Stock />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.children.length).toBe(1);
  });

  test('renders in successfully submitted state', async () => {
    const tree = renderer.create(<Stock />);

    await renderer.act(async () => {
      fetch.mockResponseOnce(JSON.stringify({ data: '12345' }));

      const input = tree.root.find(el => el.type === 'input');
      input.props.onChange({
        currentTarget: { value: dataset }
      });
      const button = tree.root.find(el => el.type === 'button');
      await button.props.onClick();

      const chartWrapper = tree.root.findAllByType('div')[1]

      expect(tree.toJSON()).toMatchSnapshot();
      expect(chartWrapper.props.className.includes('show')).toBe(true);
    });
  });

  test('renders in error state', async () => {
    const tree = renderer.create(<Stock />);

    await renderer.act(async () => {
      fetch.mockRejectOnce();

      const input = tree.root.find(el => el.type === 'input');
      input.props.onChange({
        currentTarget: { value: dataset }
      });
      const button = tree.root.find(el => el.type === 'button');
      await button.props.onClick();

      const chartWrapper = tree.root.findAllByType('div')[1]

      const error = tree.root.findAllByType('h1')[1];

      expect(tree.toJSON()).toMatchSnapshot();
      expect(chartWrapper.props.className.includes('show')).toBe(false);
      expect(error.props.className.includes('error-header')).toBe(true)
    });
  })
});
