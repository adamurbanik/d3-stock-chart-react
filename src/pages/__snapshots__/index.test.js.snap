// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stock /> renders correctly 1`] = `
<Stock
  d3Service={
    D3Service {
      "applyChartData": [MockFunction],
      "applyStockData": [MockFunction],
      "brushed": [MockFunction],
      "prepareChart": [MockFunction],
      "prepareStock": [MockFunction],
      "removeAll": [MockFunction],
      "removeChart": [MockFunction],
      "removeStock": [MockFunction],
      "zoomed": [MockFunction],
    }
  }
/>
`;

exports[`<Stock /> renders in error state 1`] = `
<main
  className="jsx-2153476396"
  id="main"
>
  <header
    className="jsx-2367272142 header-wrapper"
  >
    <h1
      className="jsx-2367272142"
    >
      Stock Chart
    </h1>
    <h2
      className="jsx-2367272142"
    >
      Select options to see chart data
    </h2>
  </header>
  <div
    className="jsx-44534442 container"
  >
    <form
      action="#"
      className="jsx-3112589256 search-form"
    >
      <input
        className="jsx-3112589256 search__input"
        name="search__input"
        onChange={[Function]}
        placeholder="Enter stock symbol i.e. AAPL, FB"
        type="text"
      />
      <button
        className="jsx-3112589256 search__button show"
        onClick={[Function]}
      >
        <svg
          className="jsx-3112589256 search__icon"
        >
          <use
            className="jsx-3112589256"
            xlinkHref="app/img/sprite.svg#icon-magnifying-glass"
          />
        </svg>
      </button>
    </form>
    <div
      className="jsx-1657759666 error-container"
    >
      <h1
        className="jsx-1657759666 error-header"
      >
        Sorry
      </h1>
      <p
        className="jsx-1657759666 error-message"
      >
        This data could not be downloaded. Please try again with a different stock code
         
      </p>
      <a
        className="jsx-1657759666"
        href="/"
      >
        Go back
      </a>
    </div>
    <div
      className="jsx-44534442 chart-wrapper"
    >
      <svg
        className="jsx-44534442 svg-stock"
        height="500"
        width="960"
      />
      <svg
        className="jsx-44534442 svg-chart"
        height="500"
        width="960"
      />
    </div>
  </div>
</main>
`;

exports[`<Stock /> renders in successfully submitted state 1`] = `
<main
  className="jsx-2153476396"
  id="main"
>
  <header
    className="jsx-2367272142 header-wrapper"
  >
    <h1
      className="jsx-2367272142"
    >
      Stock Chart
    </h1>
    <h2
      className="jsx-2367272142"
    >
      Select options to see chart data
    </h2>
  </header>
  <div
    className="jsx-44534442 container"
  >
    <form
      action="#"
      className="jsx-3112589256 search-form"
    >
      <input
        className="jsx-3112589256 search__input"
        name="search__input"
        onChange={[Function]}
        placeholder="Enter stock symbol i.e. AAPL, FB"
        type="text"
      />
      <button
        className="jsx-3112589256 search__button show"
        onClick={[Function]}
      >
        <svg
          className="jsx-3112589256 search__icon"
        >
          <use
            className="jsx-3112589256"
            xlinkHref="app/img/sprite.svg#icon-magnifying-glass"
          />
        </svg>
      </button>
    </form>
    <div
      className="jsx-44534442 chart-wrapper show"
    >
      <svg
        className="jsx-44534442 svg-stock"
        height="500"
        width="960"
      />
      <svg
        className="jsx-44534442 svg-chart"
        height="500"
        width="960"
      />
    </div>
  </div>
</main>
`;
